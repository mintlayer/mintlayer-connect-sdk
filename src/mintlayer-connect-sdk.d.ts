interface MintlayerClient {
  isMintlayer: true;

  setNetwork: (net: 'testnet' | 'mainnet') => void;
  getNetwork: () => 'testnet' | 'mainnet';

  request: (options: { method: string; params?: Record<string, any> }) => Promise<any>;
  connect: () => Promise<string[]>;

  /** Returns the extended public key (xPub). Warning: Exposes all derived addresses. */
  getXPub: () => Promise<string>;

  getAddresses: () => Promise<string[]>;
  getBalance: () => Promise<string>;
  getDelegations: () => Promise<any[]>;

  buildTransaction: (options: {
    type?: 'Transfer' | 'CreateDelegationId' | 'IssueToken';
    params: Record<string, any>;
  }) => Promise<any>;

  transfer: (options: { to: string; amount: string }) => Promise<any>;
  delegate: (options: { poolId: string; amount?: string }) => Promise<any>;
  issueToken: (options: { tokenData: Record<string, any>; authority: string }) => Promise<any>;
  signTransaction: (tx: any) => Promise<any>;

  on: (eventName: string, callback: (data: any) => void) => void;
}

declare global {
  interface Window {
    mintlayer: MintlayerClient;
  }
}

export default MintlayerClient;
