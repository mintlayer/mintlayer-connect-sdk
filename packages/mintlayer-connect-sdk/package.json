{
  "name": "@mintlayer/mintlayer-connect-sdk",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "build": "mkdir -p dist && cp src/mintlayer-connect-sdk.js dist/",
    "serve": "npx http-server . -p 8080",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
