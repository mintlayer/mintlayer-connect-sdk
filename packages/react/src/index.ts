export { MintlayerProvider, useMintlayer } from './context/MintlayerProvider';
