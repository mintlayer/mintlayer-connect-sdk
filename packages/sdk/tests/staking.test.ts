import { Client } from '../src/mintlayer-connect-sdk';
import fetchMock from 'jest-fetch-mock';

import { addresses, utxos } from './__mocks__/accounts/account_01'
import { pool_01 } from './__mocks__/pools/pool_01';

beforeEach(() => {
  fetchMock.resetMocks();

  // эмуляция window.mojito
  (window as any).mojito = {
    isExtension: true,
    connect: jest.fn().mockResolvedValue(addresses),
    restore: jest.fn().mockResolvedValue(addresses),
    disconnect: jest.fn().mockResolvedValue(undefined),
    request: jest.fn().mockResolvedValue('signed-transaction'),
  };

  fetchMock.doMock();

  fetchMock.mockResponse(async req => {
    const url = req.url;

    if (url.endsWith('/chain/tip')) {
      return JSON.stringify({ height: 200000 });
    }

    if (url.includes('/pool/')) {
      const poolId = url.split('/pool/')[1].split('/delegations')[0];
      if(poolId === 'tpool1tl784md209n53kuuwqxu68zav5lu5pdg8ca7kuhs6jg5lw24827q6qgxkc') {
        return JSON.stringify([
          {
            "balance": {
              "atoms": "0",
              "decimal": "0"
            },
            "creation_block_height": 195930,
            "delegation_id": "tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w",
            "next_nonce": 0,
            "spend_destination": "tmt1q86huq7e03hmk6wj8sf7hezqgnshhtwy6s8gz3ur"
          }
        ]);
      }
      if(poolId === 'tpool1tl784md209n53kuuwqxu68zav5lu5pdg8ca7kuhs6jg5lw24827q6qgxka') {
        return JSON.stringify([
          {
            "balance": {
              "atoms": "0",
              "decimal": "0"
            },
            "creation_block_height": 195930,
            "delegation_id": "tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w",
            "next_nonce": 0,
            "spend_destination": "tmt1q86huq7e03hmk6wj8sf7hezqgnshhtwy6s8gz3ut" // not belonging to the user
          }
        ]);
      }
      if(poolId === 'wrong_pool_id') {
        return {
          body: JSON.stringify(JSON.stringify({"error":"Invalid pool Id"})),
          status: 400,
          headers: {
            'Content-Type': 'application/json'
          }
        };
      }
    }

    if (url.includes('/token/')) {
      const tokenId = url.split('/token/').pop();
      if (tokenId === 'tmltk1jzgup986mh3x9n5024svm4wtuf2qp5vedlgy5632wah0pjffwhpqgsvmuq') {
        return JSON.stringify({
          "authority": "tmt1qyjlh9w9t7qwx7cawlqz6rqwapflsvm3dulgmxyx",
          "circulating_supply": {
            "atoms": "209000000000",
            "decimal": "2090"
          },
          "frozen": false,
          "is_locked": false,
          "is_token_freezable": true,
          "is_token_unfreezable": null,
          "metadata_uri": {
            "hex": "697066733a2f2f516d4578616d706c6548617368313233",
            "string": "ipfs://QmExampleHash123"
          },
          "next_nonce": 7,
          "number_of_decimals": 8,
          "token_ticker": {
            "hex": "58595a32",
            "string": "XYZ2"
          },
          "total_supply": {
            "Fixed": {
              "atoms": "100000000000000"
            }
          }
        });
      }
      if (tokenId === 'tmltk17jgtcm3gc8fne3su8s96gwj0yw8k2khx3fglfe8mz72jhygemgnqm57l7l') {
        return JSON.stringify({
          "authority": "tmt1qyjlh9w9t7qwx7cawlqz6rqwapflsvm3dulgmxyx",
          "circulating_supply": {
            "atoms": "209000000000",
            "decimal": "2090"
          },
          "frozen": false,
          "is_locked": false,
          "is_token_freezable": true,
          "is_token_unfreezable": null,
          "metadata_uri": {
            "hex": "697066733a2f2f516d4578616d706c6548617368313233",
            "string": "ipfs://QmExampleHash123"
          },
          "next_nonce": 7,
          "number_of_decimals": 11,
          "token_ticker": {
            "hex": "58595a32",
            "string": "XYZ2"
          },
          "total_supply": {
            "Fixed": {
              "atoms": "100000000000000"
            }
          }
        });
      }
    }

    if (url.includes('/delegation/')) {
      const delegationId = url.split('/delegation/').pop();
      if(delegationId === 'tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w') {
        return JSON.stringify({
          "balance": {
            "atoms": "0",
            "decimal": "0"
          },
          "creation_block_height": 195930,
          "next_nonce": 0,
          "pool_id": "tpool1tl784md209n53kuuwqxu68zav5lu5pdg8ca7kuhs6jg5lw24827q6qgxkc",
          "spend_destination": "tmt1q86huq7e03hmk6wj8sf7hezqgnshhtwy6s8gz3ur"
        })
      }
    }

    if(url.endsWith('/account')) {
      return {
        body: JSON.stringify({
          utxos: utxos,
        }),
      };
    }

    console.warn('No mock for:', url);
    return JSON.stringify({ error: 'No mock defined' });
  });
});

test('create delegation - snapshot', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  const spy = jest.spyOn(Client.prototype as any, 'buildTransaction');

  await client.connect();

  await client.delegationCreate({
    pool_id: 'tpool1dwpe7zy0mhagnwl36ywt5q20xxvu5dwmph4z6q8sc0a3srz5h8jqr0r2yg',
    destination: 'tmt1qyrjfd5e3nref7zga24jcthffahjwyg3csxu3xgc',
  });

  const result = await spy.mock.results[0]?.value;

  expect(result).toMatchSnapshot();
  spy.mockRestore();
});

test('delegate staking - snaphsot', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  const spy = jest.spyOn(Client.prototype as any, 'buildTransaction');

  await client.connect();

  await client.delegationStake({
    delegation_id: 'tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w',
    amount: 10,
  });

  const result = await spy.mock.results[0]?.value;

  expect(result).toMatchSnapshot();
  spy.mockRestore();
})

test('delegate staking providing only pool_id - snapshot', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  const spy = jest.spyOn(Client.prototype as any, 'buildTransaction');

  await client.connect();

  await client.delegationStake({
    pool_id: 'tpool1tl784md209n53kuuwqxu68zav5lu5pdg8ca7kuhs6jg5lw24827q6qgxkc',
    amount: 10,
  });

  const result = await spy.mock.results[0]?.value;

  expect(result).toMatchSnapshot();
  spy.mockRestore();
})

test('delegate staking providing only wrong pool_id', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  await client.connect();

  await expect(client.delegationStake({
    pool_id: 'wrong_pool_id',
    amount: 10,
  })).rejects.toThrow('Failed to fetch delegation id')
})

test('delegate staking providing only pool_id user not delegated to', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  await client.connect();

  await expect(client.delegationStake({
    pool_id: 'tpool1tl784md209n53kuuwqxu68zav5lu5pdg8ca7kuhs6jg5lw24827q6qgxka', // pool_id user not delegated to
    amount: 10,
  })).rejects.toThrow('No delegation id found for the given pool id')
})

test('staking withdraw - snapshot', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  const spy = jest.spyOn(Client.prototype as any, 'buildTransaction');

  await client.connect();

  await client.delegationWithdraw({
    delegation_id: 'tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w',
    amount: 10,
  });

  const result = await spy.mock.results[0]?.value;

  expect(result).toMatchSnapshot();
  spy.mockRestore();
})
