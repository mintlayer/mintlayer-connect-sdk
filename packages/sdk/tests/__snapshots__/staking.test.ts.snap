// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create delegation - snapshot 1`] = `
{
  "BINRepresentation": {
    "inputs": [
      Uint8Array [
        0,
        0,
        175,
        59,
        95,
        173,
        32,
        246,
        249,
        126,
        178,
        16,
        147,
        78,
        148,
        33,
        118,
        247,
        247,
        208,
        247,
        4,
        35,
        89,
        6,
        89,
        238,
        14,
        2,
        23,
        5,
        58,
        124,
        171,
        1,
        0,
        0,
        0,
      ],
    ],
    "outputs": [
      Uint8Array [
        5,
        1,
        7,
        36,
        182,
        153,
        140,
        199,
        148,
        248,
        72,
        234,
        171,
        44,
        46,
        233,
        79,
        111,
        39,
        17,
        17,
        196,
        107,
        131,
        159,
        8,
        143,
        221,
        250,
        137,
        187,
        241,
        209,
        28,
        186,
        1,
        79,
        49,
        153,
        202,
        53,
        219,
        13,
        234,
        45,
        0,
        240,
        195,
        251,
        24,
        12,
        84,
        185,
        228,
      ],
      Uint8Array [
        0,
        0,
        15,
        224,
        13,
        40,
        20,
        9,
        13,
        6,
        1,
        134,
        236,
        69,
        4,
        87,
        208,
        154,
        217,
        128,
        115,
        147,
        216,
        156,
        236,
        156,
        113,
        214,
        32,
        96,
        33,
      ],
    ],
    "transactionsize": 231,
  },
  "HEXRepresentation_unsigned": "0100040000af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab010000000805010724b6998cc794f848eaab2c2ee94f6f271111c46b839f088fddfa89bbf1d11cba014f3199ca35db0dea2d00f0c3fb180c54b9e400000fe00d2814090d060186ec450457d09ad9807393d89cec9c71d6206021",
  "JSONRepresentation": {
    "fee": {
      "atoms": "23100000000",
      "decimal": "0.231",
    },
    "id": "a1795b82f523caeafb92a50372d7b6cfb82e370831391e1172b798800e513be4",
    "inputs": [
      {
        "input": {
          "index": 1,
          "input_type": "UTXO",
          "source_id": "af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab",
          "source_type": "Transaction",
        },
        "utxo": {
          "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "1703205604300000",
              "decimal": "17032.056043",
            },
            "type": "Coin",
          },
        },
      },
    ],
    "outputs": [
      {
        "destination": "tmt1qyrjfd5e3nref7zga24jcthffahjwyg3csxu3xgc",
        "pool_id": "tpool1dwpe7zy0mhagnwl36ywt5q20xxvu5dwmph4z6q8sc0a3srz5h8jqr0r2yg",
        "type": "CreateDelegationId",
      },
      {
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
        "type": "Transfer",
        "value": {
          "amount": {
            "atoms": "1703182504300000",
            "decimal": "17031.825043",
          },
          "type": "Coin",
        },
      },
    ],
  },
  "transaction_id": "a1795b82f523caeafb92a50372d7b6cfb82e370831391e1172b798800e513be4",
}
`;

exports[`delegate staking - snaphsot 1`] = `
{
  "BINRepresentation": {
    "inputs": [
      Uint8Array [
        0,
        0,
        175,
        59,
        95,
        173,
        32,
        246,
        249,
        126,
        178,
        16,
        147,
        78,
        148,
        33,
        118,
        247,
        247,
        208,
        247,
        4,
        35,
        89,
        6,
        89,
        238,
        14,
        2,
        23,
        5,
        58,
        124,
        171,
        1,
        0,
        0,
        0,
      ],
    ],
    "outputs": [
      Uint8Array [
        6,
        7,
        0,
        16,
        165,
        212,
        232,
        109,
        61,
        61,
        216,
        42,
        173,
        158,
        59,
        189,
        40,
        132,
        167,
        34,
        120,
        238,
        48,
        62,
        119,
        49,
        3,
        170,
        159,
        60,
        54,
        25,
        131,
        33,
        6,
        53,
        9,
        194,
        53,
      ],
      Uint8Array [
        0,
        0,
        15,
        224,
        44,
        235,
        152,
        32,
        12,
        6,
        1,
        134,
        236,
        69,
        4,
        87,
        208,
        154,
        217,
        128,
        115,
        147,
        216,
        156,
        236,
        156,
        113,
        214,
        32,
        96,
        33,
      ],
    ],
    "transactionsize": 216,
  },
  "HEXRepresentation_unsigned": "0100040000af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab010000000806070010a5d4e86d3d3dd82aad9e3bbd2884a72278ee303e773103aa9f3c36198321063509c23500000fe02ceb98200c060186ec450457d09ad9807393d89cec9c71d6206021",
  "JSONRepresentation": {
    "fee": {
      "atoms": "21600000000",
      "decimal": "0.216",
    },
    "id": "775c3aff92a1cb017015bedd196b9dd2b12f1fd35148d4016416f8aac498f020",
    "inputs": [
      {
        "input": {
          "index": 1,
          "input_type": "UTXO",
          "source_id": "af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab",
          "source_type": "Transaction",
        },
        "utxo": {
          "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "1703205604300000",
              "decimal": "17032.056043",
            },
            "type": "Coin",
          },
        },
      },
    ],
    "outputs": [
      {
        "amount": {
          "atoms": "1000000000000",
          "decimal": "10",
        },
        "delegation_id": "tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w",
        "type": "DelegateStaking",
      },
      {
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
        "type": "Transfer",
        "value": {
          "amount": {
            "atoms": "1702184004300000",
            "decimal": "17021.840043",
          },
          "type": "Coin",
        },
      },
    ],
  },
  "transaction_id": "775c3aff92a1cb017015bedd196b9dd2b12f1fd35148d4016416f8aac498f020",
}
`;

exports[`delegate staking providing only pool_id - snapshot 1`] = `
{
  "BINRepresentation": {
    "inputs": [
      Uint8Array [
        0,
        0,
        175,
        59,
        95,
        173,
        32,
        246,
        249,
        126,
        178,
        16,
        147,
        78,
        148,
        33,
        118,
        247,
        247,
        208,
        247,
        4,
        35,
        89,
        6,
        89,
        238,
        14,
        2,
        23,
        5,
        58,
        124,
        171,
        1,
        0,
        0,
        0,
      ],
    ],
    "outputs": [
      Uint8Array [
        6,
        7,
        0,
        16,
        165,
        212,
        232,
        109,
        61,
        61,
        216,
        42,
        173,
        158,
        59,
        189,
        40,
        132,
        167,
        34,
        120,
        238,
        48,
        62,
        119,
        49,
        3,
        170,
        159,
        60,
        54,
        25,
        131,
        33,
        6,
        53,
        9,
        194,
        53,
      ],
      Uint8Array [
        0,
        0,
        15,
        224,
        44,
        235,
        152,
        32,
        12,
        6,
        1,
        134,
        236,
        69,
        4,
        87,
        208,
        154,
        217,
        128,
        115,
        147,
        216,
        156,
        236,
        156,
        113,
        214,
        32,
        96,
        33,
      ],
    ],
    "transactionsize": 216,
  },
  "HEXRepresentation_unsigned": "0100040000af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab010000000806070010a5d4e86d3d3dd82aad9e3bbd2884a72278ee303e773103aa9f3c36198321063509c23500000fe02ceb98200c060186ec450457d09ad9807393d89cec9c71d6206021",
  "JSONRepresentation": {
    "fee": {
      "atoms": "21600000000",
      "decimal": "0.216",
    },
    "id": "775c3aff92a1cb017015bedd196b9dd2b12f1fd35148d4016416f8aac498f020",
    "inputs": [
      {
        "input": {
          "index": 1,
          "input_type": "UTXO",
          "source_id": "af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab",
          "source_type": "Transaction",
        },
        "utxo": {
          "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "1703205604300000",
              "decimal": "17032.056043",
            },
            "type": "Coin",
          },
        },
      },
    ],
    "outputs": [
      {
        "amount": {
          "atoms": "1000000000000",
          "decimal": "10",
        },
        "delegation_id": "tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w",
        "type": "DelegateStaking",
      },
      {
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
        "type": "Transfer",
        "value": {
          "amount": {
            "atoms": "1702184004300000",
            "decimal": "17021.840043",
          },
          "type": "Coin",
        },
      },
    ],
  },
  "transaction_id": "775c3aff92a1cb017015bedd196b9dd2b12f1fd35148d4016416f8aac498f020",
}
`;

exports[`staking withdraw - snapshot 1`] = `
{
  "BINRepresentation": {
    "inputs": [
      Uint8Array [
        1,
        0,
        0,
        109,
        61,
        61,
        216,
        42,
        173,
        158,
        59,
        189,
        40,
        132,
        167,
        34,
        120,
        238,
        48,
        62,
        119,
        49,
        3,
        170,
        159,
        60,
        54,
        25,
        131,
        33,
        6,
        53,
        9,
        194,
        53,
        7,
        0,
        16,
        165,
        212,
        232,
      ],
    ],
    "outputs": [
      Uint8Array [
        1,
        0,
        7,
        0,
        130,
        186,
        3,
        231,
        1,
        245,
        126,
        3,
        217,
        124,
        111,
        187,
        105,
        210,
        60,
        19,
        235,
        228,
        64,
        68,
        225,
        123,
        173,
        196,
        212,
        2,
        129,
        112,
      ],
    ],
    "transactionsize": 78,
  },
  "HEXRepresentation_unsigned": "0100040100006d3d3dd82aad9e3bbd2884a72278ee303e773103aa9f3c36198321063509c235070010a5d4e8040100070082ba03e701f57e03d97c6fbb69d23c13ebe44044e17badc4d4028170",
  "JSONRepresentation": {
    "fee": {
      "atoms": "7800000000",
      "decimal": "0.078",
    },
    "id": "0d63777593b1fa2517b1efa3887629f138f987a01fd880eef96b38f5ea184859",
    "inputs": [
      {
        "input": {
          "account_type": "DelegationBalance",
          "amount": {
            "atoms": "1000000000000",
            "decimal": "10",
          },
          "delegation_id": "tdelg1d57nmkp24k0rh0fgsjnjy78wxql8wvgr420ncdsesvssvdgfcg6sx6262w",
          "input_type": "Account",
          "nonce": 0,
        },
      },
    ],
    "outputs": [
      {
        "destination": "tmt1q86huq7e03hmk6wj8sf7hezqgnshhtwy6s8gz3ur",
        "lock": {
          "content": "7200",
          "type": "ForBlockCount",
        },
        "type": "LockThenTransfer",
        "value": {
          "amount": {
            "atoms": "992200000000",
            "decimal": "9.922",
          },
          "type": "Coin",
        },
      },
    ],
  },
  "transaction_id": "0d63777593b1fa2517b1efa3887629f138f987a01fd880eef96b38f5ea184859",
}
`;
